File: src/main/java/kr/hahaha98757/zombiesaddon/features/PlayerVisibility.java
//Code in Player Visibility by tahmid-23
//Code in Show Spawn Time by Seosean

package kr.hahaha98757.zombiesaddon.features;

import kr.hahaha98757.zombiesaddon.config.Hotkeys;
import kr.hahaha98757.zombiesaddon.config.ZombiesAddonConfig;
import kr.hahaha98757.zombiesaddon.utils.Utils;
import net.minecraft.client.Minecraft;
import net.minecraft.entity.Entity;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraftforge.client.event.RenderPlayerEvent;
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent;
import net.minecraftforge.fml.common.gameevent.InputEvent;

public class PlayerVisibility {
    public static boolean playerVisibility;

    public PlayerVisibility() {
        playerVisibility = ZombiesAddonConfig.getModDefaultValues()[0];
    }

    @SubscribeEvent
    public void onRender(RenderPlayerEvent.Pre event) {
        if (Utils.isModDisable()) return;

        EntityPlayer player = event.entityPlayer;
        if (player != Minecraft.getMinecraft().thePlayer && !player.isPlayerSleeping() && withinDistance(player))
            event.setCanceled(playerVisibility);
    }

    private static boolean withinDistance(EntityPlayer other) {
        return Minecraft.getMinecraft().thePlayer.getDistanceToEntity(other) < ZombiesAddonConfig.getPVRange();
    }

    @SubscribeEvent
    public void onKeyInput(InputEvent.KeyInputEvent event) {
        if (!Hotkeys.togglePV.isPressed()) return;
        playerVisibility = !playerVisibility;
        Utils.addChat("§eToggled PV to " + (playerVisibility ? "§aon" : "§coff"));
    }

    public static boolean isPlayerInvisible(Entity entity) {
        boolean flag = !entity.isInvisible();
        boolean flag1 = !flag && !entity.isInvisibleToPlayer(Minecraft.getMinecraft().thePlayer);
        boolean flag2 = entity != Minecraft.getMinecraft().thePlayer && entity instanceof EntityPlayer && Minecraft.getMinecraft().thePlayer.getDistanceToEntity(entity) < ZombiesAddonConfig.getPVRange() * 2;
        return (flag || flag1) && (flag1 || flag2);
    }

    public static float getAlpha(Entity entity) {
        double pvRange = ZombiesAddonConfig.getPVRange();

        double distance = Minecraft.getMinecraft().thePlayer.getDistanceToEntity(entity);

        if (distance <= pvRange || distance > pvRange * 2) return 0.0F;

        return (float) ((distance - pvRange) / pvRange);
    }
}

File: src/main/kotlin/kr/hahaha98757/zombiesaddon/modules/PlayerVisibility.kt
package kr.hahaha98757.zombiesaddon.modules

import kr.hahaha98757.zombiesaddon.ZombiesAddon
import kr.hahaha98757.zombiesaddon.enums.SemiPvMode
import kr.hahaha98757.zombiesaddon.utils.addTranslatedChat
import kr.hahaha98757.zombiesaddon.utils.isDisable
import kr.hahaha98757.zombiesaddon.utils.isNotPlayZombies
import kr.hahaha98757.zombiesaddon.utils.mc
import net.minecraft.entity.Entity
import net.minecraft.entity.player.EntityPlayer
import net.minecraftforge.client.event.RenderPlayerEvent
import net.minecraftforge.fml.common.eventhandler.Event
import kotlin.math.sqrt

object PlayerVisibility: ToggleableModule("Player Visibility", ZombiesAddon.instance.config.pvDefault) {
    override fun getKeyBinding() = ZombiesAddon.instance.keyBindings.togglePv
    override fun addToggleText(enabled: Boolean) =
        addTranslatedChat("zombiesaddon.modules.general.toggled", "Player Visibility", if (enabled) "§aon" else "§coff")

    override fun onEvent(event: Event) {
        if (event !is RenderPlayerEvent.Pre) return
        if (isNotPlayZombies()) return

        val player = event.entityPlayer
        if (player == mc.thePlayer) return
        if (player.isPlayerSleeping) return
        if (PvUtils.withoutRange(player)) return
        event.isCanceled = true
    }
}

object PvUtils {
    fun isSemiPv(other: Entity): Boolean {
        if (isDisable()) return false
        if (!PlayerVisibility.isEnable()) return false
        if (!ZombiesAddon.instance.config.pvToggleSemiPv) return false
        if (isNotPlayZombies()) return false
        if (other !is EntityPlayer) return false
        if (other == mc.thePlayer) return false
        if (other.isPlayerSleeping) return false
        if (!withoutRange(other)) return false
        if (mc.thePlayer.getDistanceToEntity(other) > ZombiesAddon.instance.config.pvSemiPvRange) return false
        if (other.isInvisibleToPlayer(mc.thePlayer)) return false
        return true
    }

    /** isSemiPv가 true인 경우에만 호출해야 함. */
    fun getAlpha(other: Entity): Float {
        if (!isSemiPv(other)) return 1.0f // 보험. 원래는 호출 전에 isSemiPv로 걸러야 함.
        val mode = ZombiesAddon.instance.config.pvSemiPvMode
        if (mode == SemiPvMode.FIXED)
            return ZombiesAddon.instance.config.pvSemiPvMinAlpha.toFloat()

        val distance = mc.thePlayer.getDistanceToEntity(other)
        val pvRange = ZombiesAddon.instance.config.pvRange
        val semiPvRange = ZombiesAddon.instance.config.pvSemiPvRange
        val minAlpha = ZombiesAddon.instance.config.pvSemiPvMinAlpha
        val maxAlpha = ZombiesAddon.instance.config.pvSemiPvMaxAlpha

        val ratio = (pvRange - distance) / (pvRange - semiPvRange)
        val factor = when (mode) {
            SemiPvMode.LINEAR -> ratio
            SemiPvMode.SMOOTH -> ratio * ratio
            SemiPvMode.SHARP -> sqrt(ratio)
            SemiPvMode.FIXED -> return minAlpha.toFloat() // 도달할 수 없는 구간
        }

        return (minAlpha + (maxAlpha - minAlpha) * factor).toFloat()
    }

    fun withoutRange(other: EntityPlayer) = mc.thePlayer.getDistanceToEntity(other) > ZombiesAddon.instance.config.pvRange
}